File: mobile/src/main/java/com/siliconlabs/bledemo/Base/Activities/BaseActivity.java
package com.siliconlabs.bledemo.Base.Activities;

import android.content.DialogInterface;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.siliconlabs.bledemo.Base.Dialogs.ProgressDialogWithSpinner;
import com.siliconlabs.bledemo.R;

public abstract class BaseActivity extends AppCompatActivity {
    protected static final int MILLIS = 1000;
    private static boolean showSpinnerDialog = true;

    public enum ConnectionStatus {
        CONNECTING,
        CONNECTED,
        DISCONNECTING,
        DISCONNECTED
    }

    private ProgressDialogWithSpinner connectionStatusModalDialog;

    public static boolean isSpinnerDialogVisible() {
        return showSpinnerDialog;
    }

    public static void setShowSpinnerDialogVisibility(boolean visible) {
        BaseActivity.showSpinnerDialog = visible;
    }

    public void showModalDialog(ConnectionStatus connStat) {
        showModalDialog(connStat, null);
    }

    public void showModalDialog(final ConnectionStatus connStat, @Nullable final DialogInterface.OnCancelListener cancelListener) {
        dismissModalDialog();
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                switch (connStat) {
                    case CONNECTING:
                        connectionStatusModalDialog = new ProgressDialogWithSpinner(BaseActivity.this, "Connecting...", true, -1);
                        break;
                    case CONNECTED:
                        connectionStatusModalDialog = new ProgressDialogWithSpinner(BaseActivity.this, "Connection Successful!", false, R.drawable.ic_check);
                        break;
                    case DISCONNECTING:
                        // note that the dialog state is never shown when disconnecting from a device
                        connectionStatusModalDialog = new ProgressDialogWithSpinner(BaseActivity.this, "Disconnecting...", false, R.drawable.ic_check);
                        break;
                    case DISCONNECTED:
                        connectionStatusModalDialog = new ProgressDialogWithSpinner(BaseActivity.this, "Device Disconnected", false, R.drawable.ic_check);
                }

                if (!BaseActivity.this.isFinishing()) {
                    if (connStat == ConnectionStatus.CONNECTED || connStat == ConnectionStatus.DISCONNECTED) {
                        connectionStatusModalDialog.show(MILLIS);
                    } else {
                        connectionStatusModalDialog.show();
                    }

                    connectionStatusModalDialog.setOnCancelListener(cancelListener);
                }
            }
        });
    }

    public void dismissModalDialog() {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (connectionStatusModalDialog != null && connectionStatusModalDialog.isShowing()) {
                    connectionStatusModalDialog.dismiss();
                    connectionStatusModalDialog.clearAnimation();
                    connectionStatusModalDialog = null;
                }
            }
        });
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/Base/activities/BaseActivity.kt
package com.siliconlabs.bledemo.base.activities

import android.app.Activity
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import com.siliconlabs.bledemo.R
import com.siliconlabs.bledemo.base.dialogs.ProgressDialogWithSpinner
import com.siliconlabs.bledemo.utils.CustomToastManager
import timber.log.Timber
abstract class BaseActivity : AppCompatActivity() {
    enum class ConnectionStatus {
        CONNECTING, READING_DEVICE_STATE
    }

    private var connectionStatusModalDialog: ProgressDialogWithSpinner? = null
    private var isDialogVisible = false // Custom flag for dialog visibility

    @JvmOverloads
    fun showModalDialog(
        connStat: ConnectionStatus? = null,
        cancelListener: () -> Unit = {},
    ) {
        dismissModalDialog()
        runOnUiThread {
            val dialogMessage = when(connStat) {
                ConnectionStatus.CONNECTING ->
                    R.string.demo_connection_progress_dialog_connecting
                ConnectionStatus.READING_DEVICE_STATE ->
                    R.string.demo_connection_progress_dialog_reading_device_state
                else -> R.string.demo_connection_progress_dialog_loading
            }

            // note that the dialog state is never shown when disconnecting from a device
            connectionStatusModalDialog = ProgressDialogWithSpinner(
                caption = dialogMessage,
                onCancelAction = cancelListener,
            )

            if (!this@BaseActivity.isFinishing) {
                connectionStatusModalDialog?.show(supportFragmentManager, null)
                isDialogVisible = true
            }
        }
    }

    fun setModalDialogMessage(@StringRes message: Int) {
        runOnUiThread { connectionStatusModalDialog?.setCaption(message) }
    }

    fun dismissModalDialog() {
        runOnUiThread {
            if (isDialogVisible) {
                Timber.e("Attempting to dismiss the dialog")
                connectionStatusModalDialog?.dismiss()
                connectionStatusModalDialog = null
                isDialogVisible = false // Update the visibility flag after dismissing
            } else {
                Timber.e("Dialog is not visible, cannot dismiss")
            }

        }
    }

    fun showMessage(message: String) {
        runOnUiThread {
            //Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
            CustomToastManager.show(this@BaseActivity,message,5000)
        }
    }

    fun showMessage(stringResId: Int) {
        runOnUiThread {
            //Toast.makeText(this, stringResId, Toast.LENGTH_SHORT).show()
            CustomToastManager.show(this@BaseActivity,getString(stringResId),5000)
        }
    }

    fun showLongMessage(message: String) {
        runOnUiThread {
            //Toast.makeText(this, message, Toast.LENGTH_LONG).show()
            CustomToastManager.show(this@BaseActivity,message,5000)
        }
    }

    fun hideKeyboard() {
        val imm = getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
        var view = this.currentFocus
        //If no view currently has focus, create a new one, just so we can grab a window token from it
        if (view == null) view = View(this)
        imm.hideSoftInputFromWindow(view.windowToken, 0)
    }
}